Region/Region.cc

```diff
@@ -364,16 +364,55 @@ casacore::IPosition Region::xyShape() {
 // ***********************************
 // Region data
 
-bool Region::getData(std::vector<float>& data, casacore::SubLattice<float>& sublattice) {
+bool Region::getData(std::vector<float>& data, casacore::SubLattice<float>& sublattice, int checkPerChannels) {
     // fill data vector using region SubLattice
     bool dataOK(false);
     casacore::IPosition sublattShape = sublattice.shape();
     data.resize(sublattShape.product());
-    casacore::Array<float> tmp(sublattShape, data.data(), casacore::StorageInitPolicy::SHARE);
-    try {
-        sublattice.doGetSlice(tmp, casacore::Slicer(casacore::IPosition(sublattShape.size(), 0), sublattShape));
-        dataOK = true;
-    } catch (casacore::AipsError& err) {
+    if (checkPerChannels > 0 && sublattShape.size() > 2) { // stoppable spectral profile process
+        try {
+            casacore::IPosition start(sublattShape.size(), 0);
+            casacore::IPosition count(sublattShape);
+            int profileAxis; // profile axis index
+            size_t profileSize; // profile vector size
+            // get profile axis index and its vector size
+            for (int i=0; i<sublattShape.size(); ++i) {
+                if (count(i)>1) {
+                    profileAxis = i;
+                    profileSize = count(i);
+                }
+            }
+            size_t begin = 0; // the begin index of profile vector in each copy
+            size_t upperBound = (profileSize%checkPerChannels == 0 ? profileSize/checkPerChannels : profileSize/checkPerChannels + 1);
+            // get regional parameters setting as this moment
+            std::vector<CARTA::Point> tmp_ctrlpoints = m_ctrlpoints;
+            // get profile data section by section with a specific length (i.e., checkPerChannels)
+            for (size_t i=0; i<upperBound; ++i) {
+                if (pointsChanged(tmp_ctrlpoints)) { // check if regional parameters setting changed during this loop
+                    data.clear();
+                    return false; // stop profile process and don't send the data
+                }
+                // modify the start position for slicer
+                start(profileAxis) = i*checkPerChannels;
+                 // modify the count for slicer
+                count(profileAxis) = (checkPerChannels*(i+1) < profileSize ? checkPerChannels : profileSize - i*checkPerChannels);
+                casacore::Slicer slicer(start, count);
+                casacore::Array<float> buffer;
+                sublattice.doGetSlice(buffer, slicer);
+                std::vector<float> tmp = buffer.tovector();
+                memcpy(&data[begin], &tmp[0], tmp.size()*sizeof(float));
```

Comment:
you can copy directly from the `casacore::Array` using `casacore::Array::data()` without creating a temporary vector. In fact, you could even reuse the `data` vector, and just create a `casacore::Array` that uses the correct section of the `data` vector.

Reply:
Thanks for the suggestion! I have changed the code and use `casacore::Array::data()` to copy data.